// penet/lib/penet.js
var http = require("http"),
    fs   = require("fs"),
    util = require("./util"),
    url  = require("url"),
    core = require("./core"),
    Boot = require("./boot").Boot,
    BootCollection = require("./boot").BootCollection,
    Handle = require('./handle').Handle,
    HandleCollection = require("./handle").HandleCollection,
    doMessage = core.doMessage;


/**
 * Node - PENET
 *
 * VERSION 0.0.1
 *
 * Pent:
 *  field:
 *      encoding - global encoding
 *      boots - where to start webapp by url
 *      handles - handle data from request
 *
 *  function:
 *      addBoot - add a boot to boots
 *      addHandle - add a handle to handles
 *      static - set some static resource
 *      start - start server
 *
 * @param options
 * @constructor
 */

var Penet  = function(options){

    this.VERSION = "0.0.1";
    this.encoding = "UTF-8";
    this.boots = new BootCollection();
    this.handles = new HandleCollection();
    this.static_dir;

    this.addBoot = function(url,boot){
        this.boots[url] = boot
    }

    this.addHandle = function(method, handleName ,handle){
        this.handles[method][handleName] = handle;
    };

    this.static = function(options){
        var options = options || {};
        this.static_dir = options.dir;
    };

};

/**
 * Do Message
 * @param port
 */
Penet.prototype.start =  function(port){
    var me = this;
    var server = http.createServer(function(incomingMessage, response){
        doMessage(incomingMessage,function(params){
            fs.readFile(me.static_dir + url.parse(incomingMessage.url).pathname, {
                encoding: me.encoding
            },function(err, data){
                var template = null;
                try{
                    template = new Buffer(data).toString(me.encoding);
                }catch (err){

                }
                if(params.handle && params.method){
                    var handle = new Handle(me.handles[incomingMessage.method][params.handle]);

                    handle.incomingMessage = incomingMessage;
                    handle.static_dir = me.static_dir;
                    handle.response = response;
                    handle.template = template;

                    return handle[params.method]();
                }else if(template){
                    response.writeHead(200,{ 'Content-Type': "text/html"});
                    response.write(template,me.encoding);
                    response.end();
                }else{

                }
            });
        });
    });
    if(!port) port = 3000;
    server.listen(port);
};

module.exports = Penet;
