// penet/lib/util.js
var url = require("url");

var util = {};

util.extend = function(target, src, skipArr){
    var skipArr = typeof skipArr == "Array" ? skipArr : [];
    for(k in src){
        if (! _.contains(skipArr, k)) target[k] = src[k];
    }
};

util.clone = function(obj){
    return obj.constructor(obj.valueOf());
}


module.exports = {
    logger: {
        debug: function(message){
            console.debug(new Date().toString() + "[DEBUG] - " + (message || "undefined message"));
        },

        log: function(message){
            console.log(new Date().toString() + "[LOG] - " + (message || "undefined message"));
        },

        info: function(message){
            console.info(new Date().toString() + "[INFO] - " + (message || "undefined message"));
        },

        warn: function(message){
            console.warn(new Date().toString() + "[WARN] - " + (message || "undefined message"));
        },

        error: function(message){
            console.error(new Date().toString() + "[ERROR] - " + (message || "undefined message"));
        }
    }
};
